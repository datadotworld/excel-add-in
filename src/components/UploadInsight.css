/*
 * Copyright 2018 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
 .upload-row {
   margin: 0;
   height: 100vh;
 }

 .insight-upload {
    width: 85%;
    margin: 0 auto;
  }

 .insight-selected {
    max-width: 100%;
    max-height: 100%;
    margin: 30px 0;
}

.input-group, .dropdown-menu {
  width: 100%;
}

.insight-label {
  margin-top: 20px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.insight-info {
 float: right;
 font-weight: normal;
 color: gray;
 font-size: 12px;
}

.insight-upload-buttons {
  display: flex;
  justify-content: space-between;
  margin: 35px 0;
}

.insight-upload-button {
  width: 45.5%;
}

.insight-upload .dropdown {
  width: 100%;
}

.insight-upload .dropdown-toggle {
  color: #5d6f85;
  background-color: white;
  box-shadow: none;
  border-color: #6290c3;
}

.projects {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: #5d6f85;
  padding-left: 12px;
  padding-right: 12px;
  box-shadow: none;
  border: 1px solid #a3afbf;
  border-radius: 0;
}
