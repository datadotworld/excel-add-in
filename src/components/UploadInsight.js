/*
 * Copyright 2018 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { 
  Button,
  ControlLabel,
  FormControl,
  FormGroup,
  InputGroup,
  Row,
  DropdownButton,
  MenuItem
} from 'react-bootstrap';
import Published from './Published';

import './UploadInsight.css';
import analytics from '../analytics';


class UploadInsight extends Component {
  static propTypes = {
    getImageAndTitle: PropTypes.func,
    chart: PropTypes.string,
    title: PropTypes.string,
    uri: PropTypes.string,
    uploadChart: PropTypes.func,
    projects: PropTypes.array
  }

  state = {
    title: this.props.title || '',
    project: this.props.projects[0],
    description: '',
    uploadComplete: false,
    uri: ''
  }

  handleChange = (e) => {
    const { name, value } = e.target;

    this.setState({ [name]: value});
  }

  onSelect = (eventKey) => {
    const projects = this.props.projects;
    const selected = {
      owner: projects[eventKey].owner,
      id: projects[eventKey].id,
      title: projects[eventKey].title
    };

    this.setState({ project: selected });
  }

  upload = () => {
    analytics.track('exceladdin.create_project.ok.click');
    const { chart } = this.props;
    const { title, project, description } = this.state;
    this.props.uploadChart(chart, { title, project, description }).then(res => {
      this.setState({ uploadComplete: true, uri: res });
    });
  }

  closeClicked = () => {
    analytics.track('exceladdin.upload_insight.close.click');
    window.location.pathname = '/insights';
  }

  render() {
    const { uploadComplete, title, project, uri } = this.state;
    const { chart } = this.props;
    return (
      <Row className="upload-row">
        {!uploadComplete && <div className="insight-upload">
          <img
            className="insight-selected"
            src={`data:image/png;base64, ${chart}`}
            alt="chart"
          />
          <FormGroup>
            <ControlLabel className="insight-label">Project</ControlLabel>
            <DropdownButton
              id="projects-dropdown"
              className="projects"
              title={`${project.owner}/${project.title}`}
              onSelect={this.onSelect}
            >
              {
                this.props.projects.map((project, index) => {
                  return <MenuItem eventKey={index} key={index}>{`${project.owner}/${project.title}`}</MenuItem>
                })
              }
            </DropdownButton>
          </FormGroup>
          <FormGroup validationState={this.state.title.length > 60 ? 'error' : null}>
            <ControlLabel className="insight-label">
              Title <span className='insight-info'>Max. 60</span>
            </ControlLabel>
            <InputGroup>
              <FormControl
                onChange={this.handleChange}
                name='title'
                value={title}
                type='text' />
            </InputGroup>
          </FormGroup>
          <FormGroup>
            <ControlLabel className="insight-label">
              Add Comment <span className="insight-info">Optional</span>
            </ControlLabel>
            <InputGroup>
              <FormControl
                onChange={this.handleChange}
                name='description'
                value={this.state.description}
                componentClass="textarea"
                type='textarea' />
            </InputGroup>
          </FormGroup>
          <div className='insight-upload-buttons'>
            <Button
              onClick={this.closeClicked}
              className="insight-upload-button"
            >
              Cancel
            </Button>
            <Button
              className="insight-upload-button"
              onClick={this.upload}
              disabled={!title || title.length > 60}
              bsStyle='primary'
            >
              OK
            </Button>
          </div>
        </div>}
        {uploadComplete && <Published chart={chart} uri={uri} title={title} />}
      </Row>
    );
  }
}

export default UploadInsight;
