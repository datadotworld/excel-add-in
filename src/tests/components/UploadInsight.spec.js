/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';

import UploadInsight from '../../components/UploadInsight';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<UploadInsight projects={[{ owner: 'owner', id: "project"}]}/>, div);
});

it('renders upload insight view', () => {
  const getImageAndTitle = () => {
    return new Promise((resolve, reject) => {
      resolve({ image: 'b64string', title: 'A Title' });
    })
  }

  const selectedChart = 'b64ImageString'
  const projects = [
    { owner: 'firstOwner', id: "firstProject"},
    { owner: 'firstOwner', id: "secondProject"},
    { owner: 'secondOwner', id: "firstProject"},
    { owner: 'secondOwner', id: "secondProject"},
  ]

  expect(renderer.create(<UploadInsight
    chart={selectedChart}
    projects={projects}
  />).toJSON()).toMatchSnapshot()
});
