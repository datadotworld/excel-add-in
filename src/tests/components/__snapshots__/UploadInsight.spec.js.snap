// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders upload insight view 1`] = `
<div>
  <div
    className="center-block section-header insight-header row"
  >
    <div
      className="insight-title"
    >
      New Insighst
    </div>
  </div>
  <div
    className="upload-row row"
  >
    <div
      className="insight-upload"
    >
      <img
        alt="chart"
        className="insight-selected"
        src="data:image/png;base64, b64ImageString"
      />
      <div
        className="form-group"
      >
        <div
          className="body"
        >
          <label
            className="insight-label control-label"
            htmlFor={undefined}
          >
            Project URL:
          </label>
          <span
            className="input-group"
          >
            <div
              className="url"
            >
              <input
                className="textField form-control"
                id={undefined}
                onChange={[Function]}
                placeholder="https://data.world/"
                type="text"
                value="sampleProject"
              />
              <button
                className="browse-button btn btn-default"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Browse
              </button>
            </div>
          </span>
          <span
            className="help-block help-block"
          >
            Copy/paste the URL of a project, or click "Browse"
          </span>
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className="insight-label control-label"
          htmlFor={undefined}
        >
          Title 
          <span
            className="insight-info"
          >
            Max. 60
          </span>
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            id={undefined}
            name="title"
            onChange={[Function]}
            type="text"
            value=""
          />
        </span>
      </div>
      <div
        className="form-group"
      >
        <label
          className="insight-label control-label"
          htmlFor={undefined}
        >
          Add Comment 
          <span
            className="insight-info"
          >
            Optional
          </span>
        </label>
        <span
          className="input-group"
        >
          <textarea
            className="form-control"
            id={undefined}
            name="description"
            onChange={[Function]}
            type="textarea"
            value=""
          />
        </span>
      </div>
      <div
        className="insight-upload-buttons"
      >
        <button
          className="insight-upload-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="insight-upload-button btn btn-primary"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          OK
        </button>
      </div>
    </div>
  </div>
</div>
`;
